### PR TITLE
StepResults in SetMetadataItem and AddNodesToCLB

### LINKED ISSUE #1169: SetMetadataItemOnServer should return success, [reasons]

### PR DESCRIPTION
**Pull request by cyli**
Fixes #1169 and #1170.

Modifies the tests to test for the StepResult as the final result.

### CODE REVIEW

File: otter/convergence/steps.py [hunk 1]
```diff
@@ -192,14 +193,26 @@ class SetMetadataItemOnServer(object):
...
             append_segments('servers', self.server_id, 'metadata', self.key),
-            data={'meta': {self.key: self.value}})
+            data={'meta': {self.key: self.value}},
+            success_pred=has_code(200, 404))
+
+        def report_success(result):
+            return StepResult.SUCCESS, []
+
+        def report_failure(result):
+            return StepResult.RETRY, []
```
Comment (radix) [inline otter/convergence/steps.py]: I think you might as well put result[1] into the list here

File: otter/convergence/steps.py [hunk 2]
```diff
@@ -279,8 +292,25 @@ def as_effect(self):
...
+            If the error is a 422 - PENDING UPDATE, retry.  Otherwise, fail.
+            """
+            err_type, error, traceback = result
+            if err_type == APIError and error.code == 422:
+                # body has already become JSON
+                message = error.body.get("message", None)
+                if message and _CLB_PENDING_UPDATE_PATTERN.search(message):
+                    return StepResult.RETRY, []
+
+            return StepResult.FAILURE, []
```
Comment (radix) [inline otter/convergence/steps.py]: likewise in these spots

File: otter/convergence/steps.py (hunk 1 above)
Comment (cyli) [inline otter/convergence/steps.py]: Oh is that happening now?  cool

**Comment by manishtomar** [comment]
It will be nice if their are specific exceptions like `CLBPendingUpdate` that is moved around instead of `APIError`. Now, I am not sure if should compose of or inherit from `APIError` but will be nice if we define such exceptions and use them through out. I also feel that error parsing code should be more modular and separate from reporting code. This will be easier with more structured exceptions. Otherwise, we might end up with bits and pieces of error parsing in different step implementations.

**Comment by manishtomar** [comment]
I am not sure if it should be done in this PR but we are getting to a point where it is needed.

**Comment by cyli** [comment]
@manishtomar Good point.  I think it's worth doing in this PR.  I think I'm leaning towards ignoring APIError - besides which APIError may be getting replaced with UpstreamError?

**Comment by manishtomar** [comment]
> besides which APIError may be getting replaced with UpstreamError?

@cyli I am torn :/ https://github.com/rackerlabs/otter/issues/1184#issuecomment-78184828

File: otter/test/utils.py [hunk 3]
```diff
@@ -711,3 +712,41 @@ def __ne__(self, other):
...
+        }
+        eff = concretize_service_request(
+            authenticator=mock.MagicMock(),
+            log=mock.MagicMock(),
+            service_configs=service_configs,
+            tenant_id='000000',
+            service_request=service_request_intent)
+
+        # "authenticate"
+        eff = resolve_effect(eff, ('token', []))
```
Comment (radix) [inline otter/test/utils.py]: Makes me want to switch this to EQDispatcher :)

**Comment by radix** [comment]
LGTM!

**Comment by radix** [comment]
Oh, sorry, didn't notice you're still working on stuff here. I'll put it into in-progress.

**Comment by cyli** [comment]
As per IRL discussion, the points above should probably be addressed in an other PR.  So merged forward, and if passes, I'll merge?

**Comment by radix** [comment]
@cyli yes please, thanks :)

**Pull request merged**